File: frozen_lake.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

#define GRID_SIZE 4
#define STATE_COUNT 16
#define ACTION_COUNT 4

#define START_STATE 0
#define HOLE 2
#define GOAL 3

// Action deltas: left, down, right, up
int dr[] = {0, 1, 0, -1};
int dc[] = {-1, 0, 1, 0};

int lake[GRID_SIZE][GRID_SIZE] = {
    {0, 1, 1, 1},
    {1, 2, 1, 2},
    {1, 1, 1, 2},
    {2, 1, 1, 3}
};

double Q[STATE_COUNT][ACTION_COUNT];

int coords_to_state(int r, int c) {
    return r * GRID_SIZE + c;
}

void state_to_coords(int s, int* r, int* c) {
    *r = s / GRID_SIZE;
    *c = s % GRID_SIZE;
}

int is_terminal(int s) {
    int r, c;
    state_to_coords(s, &r, &c);
    int tile = lake[r][c];
    return tile == HOLE || tile == GOAL;
}

int step(int state, int action, int* reward, int* done) {
    int r, c;
    state_to_coords(state, &r, &c);
    int nr = r + dr[action];
    int nc = c + dc[action];

    // stay in place if out of bounds
    if (nr < 0 || nr >= GRID_SIZE || nc < 0 || nc >= GRID_SIZE)
        nr = r, nc = c;

    int new_state = coords_to_state(nr, nc);
    int tile = lake[nr][nc];

    *reward = (tile == GOAL) ? 1 : 0;
    *done = (tile == HOLE || tile == GOAL);
    return new_state;
}

int main() {
    srand(time(NULL));

    // Hyperparameters
    int episodes = 2000;
    double alpha = 0.8;
    double gamma = 0.95;
    double epsilon = 1.0;
    double epsilon_min = 0.01;
    double decay = 0.001;

    clock_t start_time = clock();

    for (int ep = 0; ep < episodes; ep++) {
        int state = START_STATE;
        int done = 0;

        while (!done) {
            int action;
            if ((double)rand() / RAND_MAX < epsilon) {
                action = rand() % ACTION_COUNT;
            } else {
                // argmax Q[state]
                action = 0;
                for (int a = 1; a < ACTION_COUNT; a++) {
                    if (Q[state][a] > Q[state][action])
                        action = a;
                }
            }

            int reward, next_done;
            int new_state = step(state, action, &reward, &next_done);

            // Q update
            double max_q_next = Q[new_state][0];
            for (int a = 1; a < ACTION_COUNT; a++) {
                if (Q[new_state][a] > max_q_next)
                    max_q_next = Q[new_state][a];
            }

            Q[state][action] = Q[state][action] * (1 - alpha) + alpha * (reward + gamma * max_q_next);
            state = new_state;
            done = next_done;
        }

        // Decay epsilon
        epsilon = fmax(epsilon_min, epsilon * exp(-decay));
    }

    clock_t end_time = clock();
    double elapsed_sec = (double)(end_time - start_time) / CLOCKS_PER_SEC;
    printf("Training completed in %.4f seconds\n\n", elapsed_sec);

    // Print policy
    char* action_chars[] = {"←", "↓", "→", "↑"};
    for (int s = 0; s < STATE_COUNT; s++) {
        int r, c;
        state_to_coords(s, &r, &c);
        int tile = lake[r][c];

        if (tile == HOLE) printf(" H ");
        else if (tile == GOAL) printf(" G ");
        else {
            int best = 0;
            for (int a = 1; a < ACTION_COUNT; a++) {
                if (Q[s][a] > Q[s][best])
                    best = a;
            }
            printf(" %s ", action_chars[best]);
        }

        if ((s + 1) % GRID_SIZE == 0) printf("\n");
    }

    return 0;
}

File: challenge_9/frozen_lake_benchmarking_exercise/frozen_lake.py
import numpy as np

# Map representation
lake = np.array([
    [0, 1, 1, 1],
    [1, 2, 1, 2],
    [1, 1, 1, 2],
    [2, 1, 1, 3]
])

state_to_coords = lambda s: (s // 4, s % 4)
coords_to_state = lambda r, c: r * 4 + c

# Action mapping: 0=left, 1=down, 2=right, 3=up
actions = [(-0, -1), (1, 0), (0, 1), (-1, 0)]
n_states = 16
n_actions = 4

# Initialize Q-table
Q = np.zeros((n_states, n_actions))

# Hyperparameters
episodes = 2000
alpha = 0.8
gamma = 0.95
epsilon = 1.0
epsilon_min = 0.01
decay = 0.001

def is_terminal(s):
    r, c = state_to_coords(s)
    return lake[r, c] in [2, 3]  # hole or goal

def step(state, action):
    r, c = state_to_coords(state)
    dr, dc = actions[action]
    nr, nc = r + dr, c + dc

    # Check bounds
    if 0 <= nr < 4 and 0 <= nc < 4:
        new_state = coords_to_state(nr, nc)
    else:
        new_state = state  # bump into wall

    tile = lake[state_to_coords(new_state)]
    reward = 1 if tile == 3 else 0
    done = tile in [2, 3]
    return new_state, reward, done

# Q-learning main loop

import time

start_time = time.time()

# Q-learning training loop
for episode in range(episodes):
    state = 0  # always start at S
    done = False

    while not done:
        if np.random.rand() < epsilon:
            action = np.random.randint(n_actions)
        else:
            action = np.argmax(Q[state])

        new_state, reward, done = step(state, action)
        Q[state, action] = Q[state, action] * (1 - alpha) + \
            alpha * (reward + gamma * np.max(Q[new_state]))
        state = new_state

    epsilon = max(epsilon_min, epsilon * np.exp(-decay))

end_time = time.time()
print(f"\nTraining completed in {end_time - start_time:.4f} seconds")

np.set_printoptions(precision=2, suppress=True)
print(Q)

# Show the final policy
actions_symbols = ['←', '↓', '→', '↑']
policy = np.full((4, 4), ' ')
for s in range(n_states):
    if is_terminal(s):
        policy[state_to_coords(s)] = 'H' if lake[state_to_coords(s)] == 2 else 'G'
    else:
        best_action = np.argmax(Q[s])
        policy[state_to_coords(s)] = actions_symbols[best_action]

print("\nFinal policy:")
for row in policy:
    print(' '.join(row))

